runtime/monitor.go

```diff
@@ -0,0 +1,334 @@
+	case logrus.FatalLevel.String():
+		logger.Level = logrus.FatalLevel
+	case logrus.PanicLevel.String():
+		logger.Level = logrus.PanicLevel
+	default:
+		panic(fmt.Sprintf("Unsupported log-level: %s", logLevel))
+	}
+
+	// Convert tags to logrus.Fields
+	fields := make(logrus.Fields, len(tags))
+	for k, v := range tags {
+		fields[k] = v
+	}
+
+	// Declare monitor so we can reference it in OnError
+	var m *monitor
+	m = &monitor{
+		Statsum: statsum.New(project, statsumConfigurer, statsum.Options{
+			OnError: func(err error) { m.ReportWarning(err) },
+		}),
+		Entry: logrus.NewEntry(logger).WithFields(fields),
+		sentry: &sentry{
+			client:  nil,
+			project: project,
+			auth:    auth,
+		},
+	}
+	return m
+}
+
+type sentry struct {
+	client     *raven.Client
+	m          sync.Mutex
+	project    string
+	expiration time.Time
+	auth       client.Auth
+}
+
+type monitor struct {
+	*statsum.Statsum
+	*logrus.Entry
+	*sentry
+	tags   map[string]string
+	prefix string
+}
+
+func (m *monitor) Time(name string, fn func()) {
+	m.Statsum.Time(name, fn)
+}
+
+func (m *monitor) ReportError(err error, message ...interface{}) string {
+	incidentID := uuid.NewRandom().String()
+	m.Entry.WithField("incidentId", incidentID).WithError(err).Error(message...)
+	m.submitError(err, fmt.Sprint(message...), raven.ERROR, incidentID)
+	return incidentID
+}
+
+func (m *monitor) ReportWarning(err error, message ...interface{}) string {
+	incidentID := uuid.NewRandom().String()
+	m.Entry.WithField("incidentId", incidentID).WithError(err).Warn(message...)
+	m.submitError(err, fmt.Sprint(message...), raven.WARNING, incidentID)
+	return incidentID
+}
+
+func (m *monitor) submitError(err error, message string, level raven.Severity, incidentID string) {
+	// Capture stack trace
+	exception := raven.NewException(err, raven.NewStacktrace(2, 5, []string{
+		"github.com/taskcluster/",
+	}))
+
+	// Create error packet
+	text := fmt.Sprintf("Error: %s\nMessage: %s", err.Error(), message)
+	packet := raven.NewPacket(text, nil, exception)
+	packet.Level = level
+	packet.EventID = incidentID
+
+	// Add incidentID and prefix to tags
+	tags := make(map[string]string, len(m.tags)+2)
+	for tag, value := range m.tags {
+		tags[tag] = value
+	}
+	tags["incidentId"] = incidentID
+	tags["prefix"] = m.prefix
+
+	// Refresh sentry DSN
+	m.sentry.m.Lock()
+	if m.sentry.expiration.Before(time.Now()) {
+		// Fetch DSN
+		res, rerr := m.sentry.auth.SentryDSN(m.sentry.project)
+		if rerr != nil {
+			m.Error("Failed to obtain sentry DSN, error: ", rerr)
+			m.Error("Failed to send error: ", err)
+			m.sentry.m.Unlock()
+			return
+		}
+		// Create or update DSN for the client
+		if m.sentry.client == nil {
+			m.sentry.client, rerr = raven.New(res.Dsn.Secret)
+		} else {
+			rerr = m.sentry.client.SetDSN(res.Dsn.Secret)
+		}
+		if rerr != nil {
+			m.Error("Obtained invalid sentry DSN, error: ", rerr)
+			m.Error("Failed to send error: ", err)
+			m.sentry.m.Unlock()
+			return
+		}
```

Comment:
No, it can't. It's subtle but `m.sentry.client, rerr = raven.New(res.Dsn.Secret)` also sets `rerr`